### PR TITLE
Bruteforce implementation for KNN.

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
All API plumbing for KNN and their dumb implementation.

- [ ] Add K-D trees.

- [ ] Add Ball trees.

- [ ] Figure out what to do during scoring if nothing in radius. (for radius search instead of K neighbors search)

https://github.com/dotnet/machinelearning/issues/1712

### CODE REVIEW

File: src/Microsoft.ML.StandardLearners/NearestNeighbor/MultiClassNearestNeighborTrainer.cs [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+            var examples = new List<VBuffer<float>>();
+            var labels = new List<int>();
+            if (context.InitialPredictor is MultiClassNearestNeighborModelParameters oldPredictor)
+            {
+                examples.AddRange(oldPredictor.Objects);
+                labels.AddRange(oldPredictor.Labels);
+            }
+            using (var ch = Host.Start("Data extraction"))
+            using (var cursor = new MultiClassLabelCursor(labelCount, data, CursOpt.Features | CursOpt.Label))
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/NearestNeighbor/MultiClassNearestNeighborTrainer.cs]: >using [](start = 12, length = 5)

I wonder if the entire training loop can be replaced with 'choose columns transformer' on top of training data to limit it to just features and label, and then saving the data view?

File: src/Microsoft.ML.StandardLearners/NearestNeighbor/NearestNeighbor.cs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+                var values = ctx.Reader.ReadFloatArray(count);
+                Host.CheckDecode(FloatUtils.IsFinite(values));
+                objects[i] = new VBuffer<float>(FeatureCount, count, values, indices);
+                objectsLabels[i] = ReadLabel(ctx.Reader);
+            }
+            Objects = objects.ToImmutableArray();
+            Labels = objectsLabels.ToImmutableArray();
+        }
+
+        private protected override void SaveCore(ModelSaveContext ctx)
```
[the remark this answers is not in the document]
Comment (justinormont) [reply src/Microsoft.ML.StandardLearners/NearestNeighbor/NearestNeighbor.cs]: Would the `IDataView` be placed in the model itself? I very much like the self-contained nature of ML.NET model zip files.

**Comment by justinormont** [comment]
First off; this is amazing and I very much look forward to having a KNN implementation in ML.NET. 
Second, can you reference the KNN issue(s) in the PR's description?

File: src/Microsoft.ML.StandardLearners/NearestNeighbor/NearestNeighbor.cs [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    public abstract class NearestNeighborBase<TArgs, TTransformer, TModel> : TrainerEstimatorBase<TTransformer, TModel>
+      where TTransformer : ISingleFeaturePredictionTransformer<TModel>
+      where TModel : IPredictor
+      where TArgs : NearestNeighborBase<TArgs, TTransformer, TModel>.ArgumentsBase, new()
+    {
+        public abstract class ArgumentsBase : LearnerInputBaseWithLabel
+        {
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Number of neighbors to use by default to determine result.", SortOrder = 50)]
+            [TGUI(SuggestedSweeps = "0,5,10,20,40")]
+            [TlcModule.SweepableDiscreteParam("K", new object[] { 0, 5, 10, 20, 40 })]
```
Comment (justinormont) [inline src/Microsoft.ML.StandardLearners/NearestNeighbor/NearestNeighbor.cs]: Is `0` an acceptable number of nearest neighbors to look at?

Assuming we're starting at `K=1`, I may recommend a log scale sweepable long param. 

```suggestion
            [TGUI(Description = "Number of neighbors to use by default to determine result", SuggestedSweeps = "1-64;log;inc:2")]
            [TlcModule.SweepableLongParamAttribute("K", 2, 64, isLogScale: true, stepSize: 2)]
```

**Comment by eerhardt** [comment]
This PR is out of date. Please resolve the conflicts and re-open if you are interesting in continuing this work.
